[RFC] Add method torch.jit._clone_module_with_class

dhruvbird opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56152 [RFC] Add method torch.jit._clone_module_with_class**

Currently, the Bundled Inputs API mutates the module in-place. It adds class methods and not instance methods. This results in a small problem that one can't re-run an already executed cell in Bento if the class has already been subject to bundled inputs.

In addition, there is no way to add bundled inputs to a module that has bundled inputs added already. This API provides a way to solve this problem as well by adding an `ignored_methods` to the call to `clone()` by allowing the implementation of bundled inputs to pass in the methods that it will add as `ignored_methods` so that when it does try to add those methods, it will be able to do so successfully.

We'll have to be careful when ignoring those methods during the call to `torch.jit._clone_module_with_class` since any bundled input that relies on a user-provided method will need to be preserved and not ignored during the clone.

Looking for feedback on whether this is an acceptable direction.

Differential Revision: [D27788394](https://our.internmc.facebook.com/intern/diff/D27788394/)

dhruvbird committed:
[RFC] Add method torch.jit._clone_module_with_class

Currently, the Bundled Inputs API mutates the module in-place. It adds class methods and not instance methods. This results in a small problem that one can't re-run an already executed cell in Bento if the class has already been subject to bundled inputs.

In addition, there is no way to add bundled inputs to a module that has bundled inputs added already. This API provides a way to solve this problem as well by adding an `ignored_methods` to the call to `clone()` by allowing the implementation of bundled inputs to pass in the methods that it will add as `ignored_methods` so that when it does try to add those methods, it will be able to do so successfully.

We'll have to be careful when ignoring those methods during the call to `torch.jit._clone_module_with_class` since any bundled input that relies on a user-provided method will need to be preserved and not ignored during the clone.

Looking for feedback on whether this is an acceptable direction.

Differential Revision: [D27788394](https://our.internmc.facebook.com/intern/diff/D27788394/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[RFC] Add method torch.jit._clone_module_with_class"


Currently, the Bundled Inputs API mutates the module in-place. It adds class methods and not instance methods. This results in a small problem that one can't re-run an already executed cell in Bento if the class has already been subject to bundled inputs.

In addition, there is no way to add bundled inputs to a module that has bundled inputs added already. This API provides a way to solve this problem as well by adding an `ignored_methods` to the call to `clone()` by allowing the implementation of bundled inputs to pass in the methods that it will add as `ignored_methods` so that when it does try to add those methods, it will be able to do so successfully.

We'll have to be careful when ignoring those methods during the call to `torch.jit._clone_module_with_class` since any bundled input that relies on a user-provided method will need to be preserved and not ignored during the clone.

Looking for feedback on whether this is an acceptable direction.

Differential Revision: [D27788394](https://our.internmc.facebook.com/intern/diff/D27788394/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[RFC] Add method torch.jit._clone_module_with_class"


Currently, the Bundled Inputs API mutates the module in-place. It adds class methods and not instance methods. This results in a small problem that one can't re-run an already executed cell in Bento if the class has already been subject to bundled inputs.

In addition, there is no way to add bundled inputs to a module that has bundled inputs added already. This API provides a way to solve this problem as well by adding an `ignored_methods` to the call to `clone()` by allowing the implementation of bundled inputs to pass in the methods that it will add as `ignored_methods` so that when it does try to add those methods, it will be able to do so successfully.

We'll have to be careful when ignoring those methods during the call to `torch.jit._clone_module_with_class` since any bundled input that relies on a user-provided method will need to be preserved and not ignored during the clone.

Looking for feedback on whether this is an acceptable direction.

Differential Revision: [D27788394](https://our.internmc.facebook.com/intern/diff/D27788394/)

[ghstack-poisoned]

gmagogsfm commented:
Hi @dhruvbird,

I just looked at the PR and I think it looks good. However, I am missing some context to fully understand if it fully supports your use case.

1) It sounds like bundled inputs are added to a module as a method. What does the method implement? Is it implementing logics to generate desired input set?
2) Current implementation requires cloning a module before changing bundled inputs. Is this ideal? In theory,  we can find a way (say wrap bundled input generation logic in a nn.Module that conforms to same TorchScript interface) to enable replacement of  bundled inputs without having to clone. Let me know this would work better.

dhruvbird commented:
> * It sounds like bundled inputs are added to a module as a method. What does the method implement? Is it implementing logics to generate desired input set?

Yes, bundled inputs method are added to the class and not the instance. @dreiss  implemented this and he'll have more context on why. I'm basically trying to clean it up so that it can be open sourced in a way that isn't a surprising API for users. https://www.internalfb.com/intern/diffusion/FBS/browsefile/master/xplat/caffe2/torch/utils/bundled_inputs.py?lines=0 is the code for bundled inputs. Basically it adds methods to expand tensors which have a constant value so that we can store them succinctly. Also see this post about why bundled inputs: https://fb.workplace.com/groups/pytorch.dev/permalink/644040799507737

> * Current implementation requires cloning a module before changing bundled inputs. Is this ideal? In theory,  we can find a way (say wrap bundled input generation logic in a nn.Module that conforms to same TorchScript interface) to enable replacement of  bundled inputs without having to clone. Let me know this would work better.

I think there was an idea to create a new class and wrap the user-provided class in this class, and have this new class implement the bundled input interface (is that what you're asking about @gmagogsfm ?). If so, I think this was considered and rejected (@dreiss  again for context). I don't recall why, but I can imagine that models can have methods other than `forward()` and one would need to do the forwarding logic for all of them.

dreiss commented:
> Current implementation requires cloning a module before changing bundled inputs. Is this ideal? In theory, we can find a way (say wrap bundled input generation logic in a nn.Module that conforms to same TorchScript interface) to enable replacement of bundled inputs without having to clone. Let me know this would work better.

I like this idea due to the encapsulation benefits, but I don't see a way to make it work with TorchScript static typing.  Let's say we have model `Trunk`, which we want to add bundled inputs to.  We create a module `TrunkInputs1` that has a method `generate_inputs`.  To `Trunk`, we add a field of type `TrunkInputs1` and a method `get_all_bundled_inputs` that calls `self._input_generator.generate_inputs()`.  Now, later, we decide that want to change the bundled inputs on our `Trunk` model, and, crucially, the source code to `generate_inputs` needs to change because we've changed the way we're doing compression.  We can't change the definition of `TrunkInputs1.generate_inputs`.  We can't replace `Trunk._input_generator` with a field of type `TrunkInputs1`.  Our only choice is to clone `Trunk`, right?  This would work if TorchScript supported some sort of polymorphism where `Trunk` could have a field of type `TrunkInputs` that had a concrete object of type `TrunkInputs1` and later be replaced with an object of type `TrunkInputs2`, but that's not possible, right?

gmagogsfm commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -719,6 +719,19 @@ void initJITBindings(PyObject* module) {
                 methods_to_optimize);
           })
+      .def(
+          "_clone_module_with_class",
```
Can we make this API name scarier, to something like "_hack_do_not_use_clone_module_with_class" to avoid accidental misuse?

gmagogsfm commented on torch/csrc/jit/api/module.h:
```diff
@@ -237,7 +238,12 @@ struct TORCH_API Module : public Object {
   // same data as the current instance but with the new type, shared ClassType
   // will be preserved as well
-  Module clone(bool inplace = false) const;
+  Module clone(
+      bool inplace = false,
+      const std::unordered_set<std::string>& ignored_methods =
+          std::unordered_set<std::string>(),
+      const std::unordered_set<std::string>& ignored_attributes =
+          std::unordered_set<std::string>()) const;
```
It might be better to create a new method instead of changing API and implementation of `clone`. In the future, when a proper solution lands we can more easily revert this work-around.

gmagogsfm commented on torch/csrc/jit/api/module.cpp:
```diff
@@ -16,6 +17,112 @@
+
+  for (Node* n : all_nodes) {
+    if (ignored_attributes.count(n->s(attr::name)) > 0) {
+      encountered_ignored_attributes.insert(n->s(attr::name));
```
This may be unnecessarily strict. As far as I can see, the goal is to ignore attributes/methods that are owned by top level module, but this check here applies to anything (more than just top level module) that can have attributes and unfortunately have attributes of ignored names.

dhruvbird replied:
@gmagogsfm I think I understand what you're saying, but I don't know what do do about it. I'll message you offline to learn some more about this and get your thoughts on possible way to address this.

dhruvbird committed:
Update on "[RFC] Add method torch.jit._clone_module_with_class"


Currently, the Bundled Inputs API mutates the module in-place. It adds class methods and not instance methods. This results in a small problem that one can't re-run an already executed cell in Bento if the class has already been subject to bundled inputs.

In addition, there is no way to add bundled inputs to a module that has bundled inputs added already. This API provides a way to solve this problem as well by adding an `ignored_methods` to the call to `clone()` by allowing the implementation of bundled inputs to pass in the methods that it will add as `ignored_methods` so that when it does try to add those methods, it will be able to do so successfully.

We'll have to be careful when ignoring those methods during the call to `torch.jit._clone_module_with_class` since any bundled input that relies on a user-provided method will need to be preserved and not ignored during the clone.

Looking for feedback on whether this is an acceptable direction.

Differential Revision: [D27788394](https://our.internmc.facebook.com/intern/diff/D27788394/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[RFC] Add method torch.jit._clone_module_with_class"


Currently, the Bundled Inputs API mutates the module in-place. It adds class methods and not instance methods. This results in a small problem that one can't re-run an already executed cell in Bento if the class has already been subject to bundled inputs.

In addition, there is no way to add bundled inputs to a module that has bundled inputs added already. This API provides a way to solve this problem as well by adding an `ignored_methods` to the call to `clone()` by allowing the implementation of bundled inputs to pass in the methods that it will add as `ignored_methods` so that when it does try to add those methods, it will be able to do so successfully.

We'll have to be careful when ignoring those methods during the call to `torch.jit._clone_module_with_class` since any bundled input that relies on a user-provided method will need to be preserved and not ignored during the clone.

Looking for feedback on whether this is an acceptable direction.

Differential Revision: [D27788394](https://our.internmc.facebook.com/intern/diff/D27788394/)

[ghstack-poisoned]

dhruvbird commented:
The main change that @ycao requested was the following (in his own words)!

So there are three types of things that can have attributes:
1) top level module: I think you'd like to ignore attributes on these
2) submodules: Unsure if you want to ignore attributes of them
3) user-defined class objects: Your design probably shouldn't check attributes on them at all.

So basically, his feedback was that the code as it was written would incorrectly fail if a sub-module method with the same name exists and is called. For example, if you see the the unit test in the diff, you'll see that the `MyInnerTestModule` class has a method named `dummy_method_not_cloned` and so does the main class `MyTestModule`. If we call `self.inner.dummy_method_not_cloned()`, and `dummy_method_not_cloned` is in the list of ignored methods, then this should not fail, but it would previously fail (before this revision).

The solution I put in place was to check the receiving object variable name and ran the check only if the object name is `self`. This is not be a 100% solution since the user can name the first argument anything (e.g. `foobar`), and that is what will show up. Maybe the right way to do this would be to get the name of the first argument of the method that is being processed. That said, naming the first argument as anything other than `self` results in a lint warning in Bento, so this is probably okay.

gmagogsfm commented:
> The main change that @gmagogsfm requested was the following (in his own words)!
> 
> So there are three types of things that can have attributes:
> 
> 1. top level module: I think you'd like to ignore attributes on these
> 2. submodules: Unsure if you want to ignore attributes of them
> 3. user-defined class objects: Your design probably shouldn't check attributes on them at all.
> 
> So basically, his feedback was that the code as it was written would incorrectly fail if a sub-module method with the same name exists and is called. For example, if you see the the unit test in the diff, you'll see that the `MyInnerTestModule` class has a method named `dummy_method_not_cloned` and so does the main class `MyTestModule`. If we call `self.inner.dummy_method_not_cloned()`, and `dummy_method_not_cloned` is in the list of ignored methods, then this should not fail, but it would previously fail (before this revision).
> 
> The solution I put in place was to check the receiving object variable name and ran the check only if the object name is `self`. This is not be a 100% solution since the user can name the first argument anything (e.g. `foobar`), and that is what will show up. Maybe the right way to do this would be to get the name of the first argument of the method that is being processed. That said, naming the first argument as anything other than `self` results in a lint warning in Bento, so this is probably okay.

I think relying on the name is probably fine given that in TS we have hard coded assumptions about first argument being named as `self`, so not only would bento give lint warning, TS would probably fail compilation too if that assumption is broken.

dhruvbird committed:
Update on "[RFC] Add method torch.jit._clone_module_with_class"


Currently, the Bundled Inputs API mutates the module in-place. It adds class methods and not instance methods. This results in a small problem that one can't re-run an already executed cell in Bento if the class has already been subject to bundled inputs.

In addition, there is no way to add bundled inputs to a module that has bundled inputs added already. This API provides a way to solve this problem as well by adding an `ignored_methods` to the call to `clone()` by allowing the implementation of bundled inputs to pass in the methods that it will add as `ignored_methods` so that when it does try to add those methods, it will be able to do so successfully.

We'll have to be careful when ignoring those methods during the call to `torch.jit._clone_module_with_class` since any bundled input that relies on a user-provided method will need to be preserved and not ignored during the clone.

Looking for feedback on whether this is an acceptable direction.

Differential Revision: [D27788394](https://our.internmc.facebook.com/intern/diff/D27788394/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[RFC] Add method torch.jit._clone_module_with_class"


Currently, the Bundled Inputs API mutates the module in-place. It adds class methods and not instance methods. This results in a small problem that one can't re-run an already executed cell in Bento if the class has already been subject to bundled inputs.

In addition, there is no way to add bundled inputs to a module that has bundled inputs added already. This API provides a way to solve this problem as well by adding an `ignored_methods` to the call to `clone()` by allowing the implementation of bundled inputs to pass in the methods that it will add as `ignored_methods` so that when it does try to add those methods, it will be able to do so successfully.

We'll have to be careful when ignoring those methods during the call to `torch.jit._clone_module_with_class` since any bundled input that relies on a user-provided method will need to be preserved and not ignored during the clone.

Looking for feedback on whether this is an acceptable direction.

Differential Revision: [D27788394](https://our.internmc.facebook.com/intern/diff/D27788394/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[RFC] Add method torch.jit._clone_module_with_class"


Currently, the Bundled Inputs API mutates the module in-place. It adds class methods and not instance methods. This results in a small problem that one can't re-run an already executed cell in Bento if the class has already been subject to bundled inputs.

In addition, there is no way to add bundled inputs to a module that has bundled inputs added already. This API provides a way to solve this problem as well by adding an `ignored_methods` to the call to `clone()` by allowing the implementation of bundled inputs to pass in the methods that it will add as `ignored_methods` so that when it does try to add those methods, it will be able to do so successfully.

We'll have to be careful when ignoring those methods during the call to `torch.jit._clone_module_with_class` since any bundled input that relies on a user-provided method will need to be preserved and not ignored during the clone.

Looking for feedback on whether this is an acceptable direction.

Differential Revision: [D27788394](https://our.internmc.facebook.com/intern/diff/D27788394/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[RFC] Add method torch.jit._clone_module_with_class"


Currently, the Bundled Inputs API mutates the module in-place. It adds class methods and not instance methods. This results in a small problem that one can't re-run an already executed cell in Bento if the class has already been subject to bundled inputs.

In addition, there is no way to add bundled inputs to a module that has bundled inputs added already. This API provides a way to solve this problem as well by adding an `ignored_methods` to the call to `clone()` by allowing the implementation of bundled inputs to pass in the methods that it will add as `ignored_methods` so that when it does try to add those methods, it will be able to do so successfully.

We'll have to be careful when ignoring those methods during the call to `torch.jit._clone_module_with_class` since any bundled input that relies on a user-provided method will need to be preserved and not ignored during the clone.

Looking for feedback on whether this is an acceptable direction.

Differential Revision: [D27788394](https://our.internmc.facebook.com/intern/diff/D27788394/)

[ghstack-poisoned]
